api/v1beta1/azuremachinetemplate_webhook.go

```diff
@@ -56,11 +56,31 @@ func (r *AzureMachineTemplate) ValidateCreate(ctx context.Context, obj runtime.O
 	t := obj.(*AzureMachineTemplate)
 	spec := t.Spec.Template.Spec
 
-	allErrs := ValidateAzureMachineSpec(spec)
```

Comment:
I'm not that familiar with the code, but it seems so since for `AzureMachineTemplate`, an empty value for `roleAssignmentName` is enforced. However, the validator for `AzureMachine` is doing exactly the opposite since it tries to parse it as an UUID in: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/4c28715a5bc78c6570153a601b16faf21ac43313/api/v1beta1/azuremachine_validation.go#L83
An empty string is not valid, and as a result, the validation for `AzureMachineTemplate` can't pass if system-assigned identity is enabled.

Reply:
> the validation for AzureMachineTemplate can't pass if system-assigned identity is enabled.

Can you please expand on why that's the case? The `roleAssignmentName` should be set to "" for templates as it's a GUID value that needs to be unique (so it can't be shared between all machines of a template). CAPZ will generate it for you on each AzureMachine so you shouldn't need to set it.

To enable system-assigned identity, simply set `identity` to `SystemAssigned` on the template, no need to set a roleAssignmentName.

```
apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
kind: AzureMachineTemplate
metadata:
  name: ${CLUSTER_NAME}-md-0
  namespace: default
spec:
  template:
    spec:
      identity: SystemAssigned
```

Reply:
Sure, I'll try to walk through the case step by step by analyzing the current code (without this PR).

1. As you point out, for `AzureMachineTemplate`, the `roleAssignmentName` should be left empty as it should be generated for each `AzureMachine`. In fact, it's validated as such here:

https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/bf5bc1c4dedd26a2e88a1af4ad3da0e91bd58c80/api/v1beta1/azuremachinetemplate_webhook.go#L61-L65

2. In addition to that, a call is made here to validate the template spec itself, using the validation code for `AzureMachine` here:

https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/bf5bc1c4dedd26a2e88a1af4ad3da0e91bd58c80/api/v1beta1/azuremachinetemplate_webhook.go#L59

3. That call, in turn, calls `ValidateSystemAssignedIdentity` that tries to parse the `roleAssignmentName` inside the template spec, and if left empty, as stated in point 1, the validation will fail here:

https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/bf5bc1c4dedd26a2e88a1af4ad3da0e91bd58c80/api/v1beta1/azuremachine_validation.go#L83

Reply:
To add more information on the issue, given this example of `AzureMachineTemplate` with `SystemAssigned` identity and not specifying `roleAssignmentName`:

```
apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
kind: AzureMachineTemplate
metadata:
  name: testma-control-plane
  namespace: default
spec:
  template:
    spec:
      identity: SystemAssigned
      dataDisks:
      - diskSizeGB: 256
        lun: 0
        nameSuffix: etcddisk
      osDisk:
        diskSizeGB: 128
        osType: Linux
      vmSize: Standard_D2s_v5
```
This is the error I got when applying it to the management cluster:
```
Error from server (Invalid): error when creating "testma.yaml": admission webhook "validation.azuremachinetemplate.infrastructure.cluster.x-k8s.io" denied the request: AzureMachineTemplate.infrastructure.cluster.x-k8s.io "testma-control-plane" is invalid: roleAssignmentName: Invalid value: "": Role assignment name must be a valid GUID. It is optional and will be auto-generated when not specified.
```

Reply:
Understood, thank you for providing more details. I'm looking at https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2111 which last changed this logic to see if this is a regression or if it never worked.

Reply:
Okay so it looks like this is not a recent regression, the commit above has the same issue. I was able to reproduce the validation error on my end.

I'm not a huge fan of stopping to use `ValidateAzureMachineSpec` for validation altogether. The main issue of forking is that we risk forgetting to add new validations in both places as we now have duplicated code.

There are a few alternatives we could go with to fix the issue:
1) don't make `ValidateSystemAssignedIdentity` fail when roleAssignmentName is `""`. Pro: this is a quick and easy fix. Con: AzureMachines are required to have a `roleAssignmentName` when identity is `SystemAssigned` so this would no longer catch those issues. It's mitigated by the fact that our defaulting webhook always defaults `roleAssignmentName` when it's empty and identity type is `SystemAssigned`.
2) extract `ValidateSystemAssignedIdentity` from `ValidateAzureMachineSpec` add it separately to AzureMachines' `ValidateCreate`, keep the logic in `azuremachinetemplate_webhook.go` the same. This is equivalent to the change you are making now but reversed (opt-out the func that doesn't apply instead of explicit opt-in all the others)

Thoughts?

Reply:
If I understand it correctly I vote for #2. Let's optimize for code maintenance instead of webhook order-of-precedence tricks.

Reply:
Sure, I opted for this approach due to me not being familiar with the code, as it looked like the safest solution. However, I see why you don’t like it and I agree with either of your suggestions. I personally would take no 2 since the code will be easier to understand since it’s not dependent on the defaulter.

Reply:
I pushed a new commit implementing approach no 2. Any suggestions?